pkg/liquidity-mining/test/GaugeWorkingBalanceHelper.test.ts

```diff
@@ -0,0 +1,339 @@
+
+describe('GaugeWorkingBalanceHelper', () => {
+  let bpt8020: Token;
+  let votingEscrow: Contract;
+  let veDelegationProxy: Contract;
+  let vault: Vault;
+  let BAL: Contract;
+  let workingBalanceHelper: Contract;
+  let admin: SignerWithAddress, user: SignerWithAddress, other: SignerWithAddress, another: SignerWithAddress;
+  let adaptor: Contract;
+  let pool: Contract;
+  let boost: Contract;
+  let gaugeController: Contract;
+  let gaugeImplementation: Contract;
+  let gaugeFactory: Contract;
+  let gauge: Contract;
+
+  before('setup signers', async () => {
+    [, admin, user, other, another] = await ethers.getSigners();
+  });
+
+  sharedBeforeEach('deploy Vault and tokens', async () => {
+    vault = await Vault.create({ admin });
+    adaptor = vault.authorizerAdaptor;
+
+    BAL = await deploy('TestBalancerToken', { args: [admin.address, 'Balancer', 'BAL'] });
+    bpt8020 = await Token.create('BAL/WETH 80/20');
+
+    pool = await deploy('TestBalancerToken', { args: [admin.address, 'Balancer-Incentivized', 'LP0'] });
+  });
+
+  sharedBeforeEach('deploy infrastructure and veDelegationProxy', async () => {
+    votingEscrow = await deploy('VotingEscrow', {
+      args: [bpt8020.address, 'Vote Escrowed Balancer BPT', 'veBAL', adaptor.address],
+    });
+
+    gaugeController = await deploy('MockGaugeController', { args: [votingEscrow.address, adaptor.address] });
+    // Type weight is ignored in the mock controller.
+    await gaugeController.add_type('Ethereum', 0);
+
+    const proxyV1 = await deploy('VotingEscrowDelegationProxy', {
+      args: [vault.address, votingEscrow.address, ZERO_ADDRESS],
+    });
+
+    boost = await deploy('VeBoostV2', { args: [proxyV1.address, votingEscrow.address] });
+
+    veDelegationProxy = await deploy('VotingEscrowDelegationProxy', {
+      args: [vault.address, votingEscrow.address, boost.address],
+    });
+  });
+
+  async function deployHelper(readTotalSupplyFromVE: boolean) {
+    workingBalanceHelper = await deploy('GaugeWorkingBalanceHelper', {
+      args: [veDelegationProxy.address, readTotalSupplyFromVE],
+    });
+  }
+
+  function itStoresParameters(readTotalSupplyFromVE: boolean) {
+    it('stores the veDelegationProxy', async () => {
+      expect(await workingBalanceHelper.getVotingEscrowDelegationProxy()).to.equal(veDelegationProxy.address);
+    });
+
+    it('stores the votingEscrow', async () => {
+      expect(await workingBalanceHelper.getVotingEscrow()).to.equal(votingEscrow.address);
+    });
+
+    it('indicates where to read supply from', async () => {
+      expect(await workingBalanceHelper.readsTotalSupplyFromVE()).to.equal(readTotalSupplyFromVE);
+    });
+  }
+
+  function itComputesWorkingBalances() {
+    const TOKENLESS_PRODUCTION = 0.4;
+    const MAX_BALANCE_RATIO = 1 / TOKENLESS_PRODUCTION;
+    const LOCK_PERIOD = 365 * DAY;
+    const stakeAmount = fp(5);
+    const bptAmount = fp(10);
```

Comment:
I was thinking whether it made sense to made some of these parameters and run the test with different values, but it might be an overkill. WDYT?

Reply:
I can randomize the lock period (has to be significant, but can't be more than 1 year). The stakeAmounts got flaky (1 wei rounding issues) when I varied them. Might be clearer to leave it as is, than introduce equalWithError?

Reply:
I think introducing `equalWithError` would be fine; we could even make the error explicitly small. I guess the 1 wei error is more than reasonable considering how we are doing the math here (mulDown).

I'd leave it up to you. Either way doesn't sound bad.